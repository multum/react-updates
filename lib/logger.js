export const title = (level, component, title) => {
  return console[level](
    `%c${component}%c > ${title}`,
    'color: green; font-weight: bold',
    'font-weight: bold'
  );
};

export const diff = (level, key, prev, next) => {
  console[level](
    `%c "${key}" %c${prev === next ? ' but not visually changed' : ''}` +
      `\n%c  - %s ` +
      `\n%c  + %s `,

    'background: white;' +
      ' color: black;' +
      ' line-height: 18px;' +
      ' margin-top: 2px',

    'background: none;' +
      ' line-height: 18px;' +
      ' color: #ababab;' +
      ' font-size: .85em',

    'background: #ffdce0;' + ' line-height: 18px;' + ' color: black',
    prev,

    'background: #cdffd8;' +
      ' line-height: 18px;' +
      ' color: black;' +
      ' margin-bottom: 2px',
    next
  );
};
