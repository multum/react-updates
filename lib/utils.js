/**
 * Copyright (c) 2020-present Andrew Vereshchak
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

export const stringify = value => {
  const type = typeof value;
  if (type === 'function' || type === 'symbol') {
    return value.toString();
  }

  if (isSet(value) || isMap(value)) {
    value = Array.from(value);
  } else if (isWeakSet(value) || isWeakMap(value)) {
    return Object.prototype.toString.call(value);
  }

  return JSON.stringify(value);
};

export const isNil = value => value === null || value === undefined;

export const isSet = target =>
  Object.prototype.toString.call(target) === '[object Set]';

export const isMap = target =>
  Object.prototype.toString.call(target) === '[object Map]';

export const isWeakSet = target =>
  Object.prototype.toString.call(target) === '[object WeakSet]';

export const isWeakMap = target =>
  Object.prototype.toString.call(target) === '[object WeakMap]';

export const getDisplayName = component => {
  const defaultDisplayName = 'Unknown';
  if (isNil(component)) {
    return defaultDisplayName;
  } else if (typeof component === 'string') {
    return component || defaultDisplayName;
  } else {
    return component.displayName || component.name || defaultDisplayName;
  }
};
