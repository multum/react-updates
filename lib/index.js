/**
 * Copyright (c) 2020-present Andrew Vereshchak
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import * as utils from './utils';

const Settings = {
  disabled: false,
};

export const setDebuggerSettings = settings => {
  return Object.assign(Settings, settings);
};

export const debugComponentUpdate = (component, prev, next) => {
  if (Settings.disabled) return;

  const displayName = utils.getDisplayName(component);

  const title = (notifier, title) => {
    notifier(
      `%c${displayName}%c > ${title}`,
      'color: green; font-weight: bold',
      'font-weight: bold'
    );
  };

  if (prev) {
    const allProperties = Object.assign({}, prev, next);
    const changed = Object.keys(allProperties).filter(
      key => !Object.is(prev[key], next[key])
    );

    if (changed.length === 0) {
      title(console.warn, 'Re-render without props changes');
      return;
    }

    const strings = [];
    const styles = [];

    changed.forEach(key => {
      const prevValue = utils.stringify(prev[key]);
      const nextValue = utils.stringify(next[key]);
      strings.push(
        `%c "${key}" %c` +
          (prevValue === nextValue ? ' but not visually changed' : '') +
          `\n%c  - ${prevValue} ` +
          `\n%c  + ${nextValue} `
      );
      styles.push(
        `
          background: white;
          color: black;
          line-height: 18px;
          border: 1px solid whitesmoke;
        `,
        `
          background: none;
          line-height: 18px;
          color: #ababab;
          font-size: .85em;
        `,
        `
          background: #ffdce0;
          line-height: 18px;
          color: black
        `,
        `
          background: #cdffd8;
          line-height: 18px;
          color: black;
          margin-bottom: 4px;
        `
      );
    });
    title(console.group, 'Re-render with props changes');
    console.info.apply(null, [strings.join('\n')].concat(styles));
    console.groupEnd();
  } else {
    title(console.info, 'Initial render');
  }
};

export const useDebugger = (component, props) => {
  if (Settings.disabled) return;

  const ref = React.useRef(null);
  React.useEffect(() => {
    debugComponentUpdate(component, ref.current, props);
    ref.current = props;
  });
};
