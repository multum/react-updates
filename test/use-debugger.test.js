import React from 'react';
import { useDebugger, setDebuggerSettings } from '../lib';
import * as utils from './utils';
import { render } from '@testing-library/react';

const DebugComponent = (props) => {
  useDebugger(DebugComponent, props);
  return <div>Child component content</div>;
};

DebugComponent.displayName = 'DebugComponent';

describe('useDebugger', () => {
  utils.mockConsole();

  it('initial render', () => {
    const styles = { width: '100%' };
    const Component = () => <DebugComponent style={styles} />;
    const { rerender } = render(<Component />);
    rerender(<Component />);
    expect(console.calls).toMatchSnapshot();
  });

  it('normal re-render', () => {
    const { rerender } = render(<DebugComponent content="foo" />);
    rerender(<DebugComponent content="bar" />);
    expect(console.calls).toMatchSnapshot();
  });

  it('redundant re-render', () => {
    const Component = () => <DebugComponent style={{ width: '100%' }} />;
    const { rerender } = render(<Component />);
    rerender(<Component />);
    expect(console.calls).toMatchSnapshot();
  });

  it('redundant re-render without change props', () => {
    const { rerender } = render(<DebugComponent />);
    rerender(<DebugComponent />);
    expect(console.calls).toMatchSnapshot();
  });

  it('disable in settings', () => {
    setDebuggerSettings({ disabled: true });
    render(<DebugComponent />);
    expect(console.calls).toEqual([]);
  });
});
