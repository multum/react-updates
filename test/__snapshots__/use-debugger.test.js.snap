// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDebugger initial render 1`] = `
Array [
  Object {
    "__": "console.info",
    "arguments": Array [
      "%cDebugComponent%c > Initial render",
      "color: green; font-weight: bold",
      "font-weight: bold",
    ],
  },
  Object {
    "__": "console.warn",
    "arguments": Array [
      "%cDebugComponent%c > Re-render without props changes",
      "color: green; font-weight: bold",
      "font-weight: bold",
    ],
  },
]
`;

exports[`useDebugger normal re-render 1`] = `
Array [
  Object {
    "__": "console.info",
    "arguments": Array [
      "%cDebugComponent%c > Initial render",
      "color: green; font-weight: bold",
      "font-weight: bold",
    ],
  },
  Object {
    "__": "console.group",
    "arguments": Array [
      "%cDebugComponent%c > Re-render with props changes",
      "color: green; font-weight: bold",
      "font-weight: bold",
    ],
  },
  Object {
    "__": "console.info",
    "arguments": Array [
      "%c \\"content\\" %c
%c  - \\"foo\\" 
%c  + \\"bar\\" ",
      "background: white; color: black; line-height: 18px; margin-top: 3px",
      "background: none; line-height: 18px; color: #ababab; font-size: .85em",
      "background: #ffdce0; line-height: 18px; color: black",
      "background: #cdffd8; line-height: 18px; color: black; margin-bottom: 4px",
    ],
  },
  Object {
    "__": "console.groupEnd",
    "arguments": Array [],
  },
]
`;

exports[`useDebugger redundant re-render 1`] = `
Array [
  Object {
    "__": "console.info",
    "arguments": Array [
      "%cDebugComponent%c > Initial render",
      "color: green; font-weight: bold",
      "font-weight: bold",
    ],
  },
  Object {
    "__": "console.group",
    "arguments": Array [
      "%cDebugComponent%c > Re-render with props changes",
      "color: green; font-weight: bold",
      "font-weight: bold",
    ],
  },
  Object {
    "__": "console.info",
    "arguments": Array [
      "%c \\"style\\" %c but not visually changed
%c  - {\\"width\\":\\"100%\\"} 
%c  + {\\"width\\":\\"100%\\"} ",
      "background: white; color: black; line-height: 18px; margin-top: 3px",
      "background: none; line-height: 18px; color: #ababab; font-size: .85em",
      "background: #ffdce0; line-height: 18px; color: black",
      "background: #cdffd8; line-height: 18px; color: black; margin-bottom: 4px",
    ],
  },
  Object {
    "__": "console.groupEnd",
    "arguments": Array [],
  },
]
`;

exports[`useDebugger redundant re-render without change props 1`] = `
Array [
  Object {
    "__": "console.info",
    "arguments": Array [
      "%cDebugComponent%c > Initial render",
      "color: green; font-weight: bold",
      "font-weight: bold",
    ],
  },
  Object {
    "__": "console.warn",
    "arguments": Array [
      "%cDebugComponent%c > Re-render without props changes",
      "color: green; font-weight: bold",
      "font-weight: bold",
    ],
  },
]
`;
